File: plugins/seurat-standard/tools/figures/generate_placeholder_figure.R
#!/usr/bin/env Rscript
# ============================================================
# generate_placeholder_figure.R — 生成 seurat-standard placeholder UMAP PNG
#
# 输出: resources/built-in-plugins/seurat-standard/tools/figures/umap_placeholder.png
#
# 设计：
#   - 256×64 灰色条带（与 scanpy-advanced 占位图风格一致）
#   - png::writePNG(png::readPNG(f), f) 重写以稳定 sha256（与 scanpy Pillow 等价）
#   - 确定性: 不依赖 random / 时间戳
#
# 运行:
#   /path/to/biof3-r-runtime/bin/Rscript generate_placeholder_figure.R
# ============================================================

if (!requireNamespace("png", quietly = TRUE)) {
  stop("[placeholder figure] R package 'png' is required")
}

# Build a minimal valid PNG by hand via png::writePNG
# Output dimensions
w <- 256
h <- 64

# Create a 4-channel RGBA matrix: h × w × 4
# Use a grey gradient with subtle banding (deterministic)
mat <- array(0L, dim = c(h, w, 4L))
for (y in seq_len(h)) {
  for (x in seq_len(w)) {
    intensity <- (x %% 64) * 3 + 128
    r <- (intensity + ((x %/% 8) %% 16)) %% 256
    g <- r
    b <- r
    a <- 255L
    mat[y, x, 1] <- r
    mat[y, x, 2] <- g
    mat[y, x, 3] <- b
    mat[y, x, 4] <- a
  }
}

script_dir <- tryCatch({
  dirname(normalizePath(sys.frame(1)$ofile))
}, error = function(e) ".")
out_path <- file.path(script_dir, "umap_placeholder.png")

# Write first time
png::writePNG(mat / 255, target = out_path)

# Re-read + re-write to canonicalize (R png::writePNG(png::readPNG(f), f) equivalent)
buf <- png::readPNG(out_path)
png::writePNG(buf, target = out_path)

sha <- digest::digest(file = out_path, algo = "sha256")
cat(sprintf("[placeholder figure] wrote %s (%d bytes, sha256-prefix=%s)\n",
            out_path,
            file.info(out_path)$size,
            substr(sha, 1, 12)))

File: plugins.dormant/celltypist/tools/figures/generate_placeholder_figure.py
"""Generate a deterministic placeholder figure for celltypist.

This produces a single PNG that is then read + re-saved via Pillow
(matching worker C §0.5.5 "PNG hash 稳定" requirement: png::writePNG(readPNG(f), f)
on the R side, Pillow Image.save on the Python side).

Output:
    resources/built-in-plugins/celltypist/tools/figures/annotation_placeholder.png

Run:
    /Applications/anaconda3/envs/biof3-py-runtime/bin/python3 tools/figures/generate_placeholder_figure.py

Idempotent: re-running produces byte-identical output (modulo the read+save
hash-rewrite loop), so the file's sha256 stays stable for Result Studio
artifact-lineage.
"""

from __future__ import annotations

import io
import struct
import zlib
from pathlib import Path

OUT_PATH = Path(__file__).parent / "annotation_placeholder.png"


def _png_chunk(tag: bytes, data: bytes) -> bytes:
    length = struct.pack(">I", len(data))
    crc = struct.pack(">I", zlib.crc32(tag + data) & 0xFFFFFFFF)
    return length + tag + data + crc


def _build_minimal_png(width: int = 256, height: int = 64) -> bytes:
    """Build a minimal valid PNG: 256×64 transparent strip with text-style grey.

    Deterministic, no timestamps, no compression metadata that varies.
    """
    sig = b"\x89PNG\r\n\x1a\n"
    ihdr_data = struct.pack(">IIBBBBB", width, height, 8, 6, 0, 0, 0)  # RGBA, 8-bit, no interlace
    ihdr = _png_chunk(b"IHDR", ihdr_data)

    # Build scanlines: filter byte 0 (None) + RGBA pixels
    raw = bytearray()
    for _ in range(height):
        raw.append(0)
        for x in range(width):
            # violet gradient (celltypist brand color theme)
            intensity = (x % 64) * 3 + 128
            r = (intensity + ((x // 8) % 16)) % 256
            g = (intensity // 2 + 64) % 256
            b = (intensity + 48) % 256
            a = 255
            raw.extend((r, g, b, a))
    idat = _png_chunk(b"IDAT", zlib.compress(bytes(raw), level=9))
    iend = _png_chunk(b"IEND", b"")
    return sig + ihdr + idat + iend


def _canonicalize_png(buf: bytes) -> bytes:
    """Read + save via Pillow so the final PNG hash matches Pillow's encoder
    (parallels R png::writePNG(png::readPNG(f), f))."""
    from PIL import Image

    img = Image.open(io.BytesIO(buf))
    img.load()
    out = io.BytesIO()
    img.save(out, format="PNG", optimize=False)
    return out.getvalue()


def main() -> None:
    raw = _build_minimal_png()
    canonical = _canonicalize_png(raw)
    OUT_PATH.write_bytes(canonical)
    import hashlib

    print(
        f"[placeholder figure] wrote {OUT_PATH} ({len(canonical)} bytes, "
        f"sha256-prefix={hashlib.sha256(canonical).hexdigest()[:12]})"
    )


if __name__ == "__main__":
    main()
